sql: constrict Min and Max times for Timestamp/TimestampTZ

otan opened the pull request:
We previously supported an arbitrary range of TIMESTAMP/TIMESTAMPTZ
values as long as it is supported by Golang. However, this results in
erroneous results for sorting, as we key indexes by Unix time, but
`time.Unix(maxInt64, 0) < time.Unix(minInt64, 0)` due to overflows in
the time library.

Furthermore, supporting these large values is incompatible with Postgres
as Postgres (and the wire protocol) only supports 8 byte timestamps,
whereas we can encode Golang style timestamps with 12 bytes.

To make this easy for ourselves, we now only support the Postgres range
of timestamps.

This introduces a backward-incompatible behavior with older versions of
CRDB, however, we expect very few people to store such huge dates.
Furthermore, these large date values are extremely buggy.

The majority of changes that are relevant are in `datum.go`, where
`MakeDTimestamp` and `MakeDTimestampTZ` now return an error if it
exceeds the given bounds. There is a `Must` intended for tests or cases
where it should be impossible to get an out of range timestamp (i.e.
when using current_timestamp or converting a HLC to a timestamp). The
rest should just be propagating errors down all the way. Note this not
affect decoded as `DecodeTableKey`.

Resolves https://github.com/cockroachdb/cockroach/issues/46973.

Release note (sql change, backward-incompatible change): We previously
supported any TIMESTAMP(TZ) representable in Go. However, this is
incompatible with the postgres protocol format, as well as resulting in
sorting errors when using values such as
`(-9223372036854775808)::TIMESTAMP`. We have changed our bounds for
timestamp to be the same as postgres - between `4714-11-24 00:00:00+00
BC` and `294276-12-31 23:59:59.999999` inclusive. Users who have
previously stored these values will be able to access them, but the WHERE
clause must use `< '4714-11-24 00:00:00+00 BC'` or
`> 294276-12-31 23:59:59.999999`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47077)
<!-- Reviewable:end -->

awoods187 commented:
>This introduces a backward-incompatible behavior with older versions of
CRDB, however, we expect very few people to store such huge dates.
Furthermore, these large date values are extremely buggy.

Will this bug fix only be in 20.2?

On Mon, Apr 6, 2020 at 12:04 PM cockroach-teamcity <notifications@github.com>
wrote:

> This change is [image: Reviewable]
> <https://reviewable.io/reviews/cockroachdb/cockroach/47077>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/47077#issuecomment-609885284>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFJ7F74PZPFLSOHLRBWUGFTRLH4R5ANCNFSM4MCMXJSQ>
> .
>


-- 
Andy Woods
Senior Product Manager, Cockroach Labs
andy@cockroachlabs.com

otan commented:
Yes. I don't anticipate people really hitting this as an issue in 20.1 or earlier though, because...well...who realistically needs to store that date haha

andy-kimball commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -3858,15 +3900,15 @@ func NewDefaultDatum(evalCtx *EvalContext, t *types.T) (d Datum, err error) {
 		return dEpochDate, nil
 	case types.TimestampFamily:
-		return dMinTimestamp, nil
+		return dZeroTimestamp, nil
```
Is `dZeroTimestamp` the same value as `dMinTimestamp`, and is it in the allowed range?

otan replied:
>  Is dZeroTimestamp the same value as dMinTimestamp?

No

> and is it in the allowed range?

Yes

andy-kimball replied:
Let's fix this immediately in the 20.1 branch, then. Otherwise, this will be a breaking change.

otan replied:
Misunderstood the question -- `dZeroTimestamp` is the same as 20.1's `dMinTimestamp`, but does not represent the minimum time that is supported.

otan commented:
@knz do you have a strong desire to review this (some of the min/max TODOs on Time types have your name on it), or should I let someone on SQL team.

knz commented:
thank you for offering; no I do not wish to review this (also OOO this week)

rohany commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -2094,11 +2108,25 @@ type DTimestamp struct {
-var dMinTimestamp = &DTimestamp{}
+// MustDTimestamp wraps MakeDTimestamp but panics if there is an error.
+// This is intended for testing applications only.
+func MustDTimestamp(t time.Time, precision time.Duration) *DTimestamp {
```
`MustDTimestamp` feels awkward. I guess we already have a function `MustBeDTimestamp`, so that's already taken. How about `MustBeValidDTimestamp`?

otan replied:
I've done `MustMakeDTimestamp`.

rohany commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -2242,39 +2270,39 @@ func (d *DTimestamp) Compare(ctx *EvalContext, other Datum) int {
+	if d.IsMin(ctx) {
+		return nil, false
+	}
 	return &DTimestamp{Time: d.Add(-time.Microsecond)}, true
```
Was this just underflowing in the meantime?

otan replied:
yep!

rohany reviewed (commented):
So we don't affect values on disk that are out of the range when decoding right? Is it possible to write a test to ensure this?

otan reviewed (commented):
see `TestDecodeTable*` for tests for encoding/decoding out of range working for the past.

otan committed:
sql: constrict Min and Max times for Timestamp/TimestampTZ

We previously supported an arbitrary range of TIMESTAMP/TIMESTAMPTZ
values as long as it is supported by Golang. However, this results in
erroneous results for sorting, as we key indexes by Unix time, but
`time.Unix(maxInt64, 0) < time.Unix(minInt64, 0)` due to overflows in
the time library.

Furthermore, supporting these large values is incompatible with Postgres
as Postgres (and the wire protocol) only supports 8 byte timestamps,
whereas we can encode Golang style timestamps with 12 bytes.

To make this easy for ourselves, we now only support the Postgres range
of timestamps.

This introduces a backward-incompatible behavior with older versions of
CRDB, however, we expect very few people to store such huge dates.
Furthermore, these large date values are extremely buggy.

The majority of changes that are relevant are in `datum.go`, where
`MakeDTimestamp` and `MakeDTimestampTZ` now return an error if it
exceeds the given bounds. There is a `Must` intended for tests or cases
where it should be impossible to get an out of range timestamp (i.e.
when using current_timestamp or converting a HLC to a timestamp). The
rest should just be propagating errors down all the way. Note this not
affect decoded as `DecodeTableKey`.

Release note (sql change, backward-incompatible change): We previously
supported any TIMESTAMP(TZ) representable in Go. However, this is
incompatible with the postgres protocol format, as well as resulting in
sorting errors when using values such as
`(-9223372036854775808)::TIMESTAMP`. We have changed our bounds for
timestamp to be the same as postgres - between `4714-11-24 00:00:00+00
BC` and `294276-12-31 23:59:59.999999` inclusive. Users who have
previously stored these values will be able to access them, but the WHERE
clause must use `< '4714-11-24 00:00:00+00 BC'` or
`> 294276-12-31 23:59:59.999999`.

rohany commented:
LGTM -- sorry about the delay... I thought I accepted it but never actually commented

otan commented:
ty!

bors r=rohany
